include/circt/Dialect/HW/HWSymCache.h

```diff
@@ -6,100 +6,71 @@
     bool hasPort() const { return port != ~0ULL; }
-    mlir::Operation *getOp() const { return op; }
     size_t getPort() const { return port; }
+    mlir::Operation *getOp() const { return op; }
 
   private:
     mlir::Operation *op;
     size_t port;
   };
 
-  /// In the building phase, add symbols.
-  void addDefinition(mlir::StringAttr symbol, mlir::Operation *op) {
-    assert(!isFrozen && "cannot mutate a frozen cache");
-    symbolCache.try_emplace(symbol, op, ~0ULL);
-  }
-
   // Add inner names, which might be ports
-  void addDefinition(mlir::StringAttr modSymbol, mlir::StringAttr name,
-                     mlir::Operation *op, size_t port = ~0ULL) {
-    assert(!isFrozen && "cannot mutate a frozen cache");
+  void addInnerDefinition(mlir::StringAttr modSymbol, mlir::StringAttr name,
+                          mlir::Operation *op, size_t port = ~0ULL) {
     auto key = InnerRefAttr::get(modSymbol, name);
     symbolCache.try_emplace(key, op, port);
   }
 
-  /// Populate the symbol cache with all symbol-defining operations within the
-  /// 'top' operation.
-  void addDefinitions(mlir::Operation *top) {
-    for (auto &region : top->getRegions())
-      for (auto &block : region.getBlocks())
-        for (mlir::Operation &op : block)
-          if (auto symOp = llvm::dyn_cast<mlir::SymbolOpInterface>(op))
-            if (auto name = symOp.getNameAttr())
-              addDefinition(name, symOp);
-  }
-
-  // Add inner names, which might be ports
-  void addDefinition(InnerRefAttr name, mlir::Operation *op,
-                     size_t port = ~0ULL) {
+  void addDefinition(mlir::Attribute key, mlir::Operation *op) override {
     assert(!isFrozen && "cannot mutate a frozen cache");
-    symbolCache.try_emplace(name, op, port);
+    symbolCache.try_emplace(key, op);
   }
 
-  /// Mark the cache as frozen, which allows it to be shared across threads.
-  void freeze() { isFrozen = true; }
-
-  mlir::Operation *getDefinition(mlir::StringAttr symbol) const {
-    return lookup(symbol);
+  HWSymbolCache::Item getInnerDefinition(mlir::StringAttr modSymbol,
+                                         mlir::StringAttr name) const {
+    return lookupInner(InnerRefAttr::get(modSymbol, name));
   }
 
-  mlir::Operation *getDefinition(mlir::FlatSymbolRefAttr symbol) const {
-    return lookup(symbol.getAttr());
+  HWSymbolCache::Item getInnerDefinition(InnerRefAttr inner) const {
+    return lookupInner(inner);
   }
 
-  Item getDefinition(mlir::StringAttr modSymbol, mlir::StringAttr name) const {
-    return lookup(InnerRefAttr::get(modSymbol, name));
-  }
-
-  Item getDefinition(InnerRefAttr name) const { return lookup(name); }
-
-protected:
-  bool isFrozen = false;
+  /// Mark the cache as frozen, which allows it to be shared across threads.
+  void freeze() { isFrozen = true; }
 
 private:
-  Item lookup(InnerRefAttr attr) const {
+  Item lookupInner(InnerRefAttr attr) const {
```

Comment:
Why go through these `lookup`s? Why not just put them in the `getDefinitions` method bodies?

Reply:
We could, but then we'd have code duplication in the two `getInnerDefinition` methods.

Reply:
Sorry, I meant to leave this comment in SymbolCache.h on the non-`Inner` variant.

Reply:
`lookup` is used through the `SymbolCacheBase` to allow for the implementing class to provide a single implementation for attribute-to-operation lookup. In this case, this is further used to add additional checks for the HWSymCache on the frozen state of the cache, even when accessed through `SymbolCacheBase`.

Reply:
I get that, but why not use `getDefinition` instead of `lookup`? The first variant is just a passthrough to `lookup`, so why not cut of the middleman? Similar to `addDefinition` -- it's a public method which is completely virtual which all implementing classes must override. Is it a C++ issue with overloads not playing nicely with virtual methods?

Reply:
Moved the lookup into `getDefinition`; initially there was an issue wrt. overload resolution between 
```
mlir::Operation *getDefinition(mlir::Attribute attr) const
mlir::Operation *getDefinition(mlir::FlatSymbolRefAttr symbol) const
```
but, this is fixed with pulling in the 2nd definition through `using SymbolCacheBase::getDefinition`.

Reply:
Great!